### PR TITLE
Wrap nn.Linear LoRA path with LoRALinear (to_wrap/adapter)

### PR DESCRIPTION
**Pull request by HollowMan6**
# What does this PR do ?

To make PEFT bridge / LoRA merge work smoothly, we need to align the simple `nn.Linear` LoRA code path with the Megatron parallel path by adopting the `to_wrap`/`adapter` wrapper pattern: plain linears are now wrapped as `LoRALinear(module, LinearAdapter(...))` instead of returning a self-contained fused `LinearAdapter`.

# Changelog

- **`LinearAdapter` (`peft/lora_layers.py`)** — refactored from an `nn.Linear` subclass that copied the base weight and fused base+LoRA in one `forward` into a delta-only `nn.Module` (mirrors `ParallelLinearAdapter`'s role): it holds only `linear_in`/`linear_out` and returns the scaled LoRA delta. No base weight/bias copy.
- **Simple path (`peft/lora.py`, `peft/canonical_lora.py`)** — `LoRA.transform`/`CanonicalLoRA.transform` now build `LinearAdapter(module, ...)` and wrap it as `LoRALinear(module, adapter)`; `LinearAdapter` removed from the already-transformed skip-tuples (it is now a sub-module, not a top-level wrapper).
- **`LoRALinear` / `base_linear_forward` (`peft/adapter_wrapper.py`, `peft/lora_layers.py`)** — generalized to handle plain `nn.Linear`'s bare-tensor return (Megatron modules return tuples). `base_linear_forward` branches on `isinstance(result, torch.Tensor)` before the tuple unpacking and records the return shape via `_base_returns_tuple`; `LoRALinear.forward` returns a bare tensor for plain `nn.Linear` (drop-in replacement) and `(out, bias)` for Megatron modules. Non-tensor, non-tuple returns still raise `AssertionError`.
- **Checkpoint keys** — adapter delta now stored under the `adapter.` prefix (`adapter.linear_in.weight`, `adapter.linear_out.weight`) via `AdapterWrapper.state_dict`, consistent with the parallel path. **Breaking:** existing simple-linear LoRA checkpoints (flat `linear_in.weight`/`linear_out.weight`) are not forward-compatible and must be re-saved.
- **Guards for plain `nn.Linear`**:
  - `AdapterWrapper.sharded_state_dict` guards `hasattr(to_wrap/adapter, "sharded_state_dict")` so it no longer crashes on plain modules.
  - `models/conversion/peft_bridge.py` short-circuits `LinearAdapter` to default parallelism attributes (`input_is_parallel=False`, `base_linear_is_parallel=False`, `requires_expert_splits=False`) instead of calling `get_adapter_attributes_from_linear(to_wrap)`, which requires `to_wrap.config` that plain `nn.Linear` lacks.
- **Tests** — updated `test_adapter_wrapper.py`, `test_lora_layers.py`, `test_lora.py`, `test_canonical_lora.py`, `test_moonlight_recipes.py` for the new `isinstance(..., LoRALinear)` wrapper type, `.adapter.*` attribute access, and `adapter.`-prefixed state-dict keys. Added `test_base_linear_forward_tensor_return` and restored `test_base_linear_forward_invalid_return` for the malformed-return guard.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci) in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: src/megatron/bridge/peft/adapter_wrapper.py [hunk 1]
```diff
@@ -140,13 +141,24 @@ def base_linear_forward(
             4. both: (out, bias, ln_out)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply src/megatron/bridge/peft/adapter_wrapper.py:169]: addressed

File: tests/unit_tests/peft/test_lora.py [hunk 2]
```diff
@@ -247,10 +247,10 @@ def test_lora_transform_simple_model(self):
         transformed_model = lora(model, training=True)
 
         # Check that target modules were transformed to LinearAdapter
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply tests/unit_tests/peft/test_lora.py:249]: addressed

File: tests/unit_tests/peft/test_canonical_lora.py [hunk 3]
```diff
@@ -284,8 +284,8 @@ def test_canonical_lora_transform_simple_model(self):
         transformed_model = lora(model, training=True)
 
         # Check that target modules were transformed to LinearAdapter (simple nn.Linear case)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply tests/unit_tests/peft/test_canonical_lora.py:286]: addressed

File: src/megatron/bridge/peft/adapter_wrapper.py [hunk 4]
```diff
@@ -213,10 +225,16 @@ def sharded_state_dict(
...
-        sharded_state_dict.update(
-            self.adapter.sharded_state_dict(
-                f"{prefix}adapter.", sharded_offsets, metadata, **adapter_sharded_state_dict_kwargs
+        # The wrapped module may be a plain nn.Linear (simple, non-parallel path) which
+        # does not implement distributed sharding; fall back gracefully in that case.
+        if hasattr(self.to_wrap, "sharded_state_dict"):
+            sharded_state_dict.update(self.to_wrap.sharded_state_dict(prefix, sharded_offsets, metadata))
+        # Likewise, a delta-only LinearAdapter has no sharded layout; only parallel adapters
+        # carry distributed-sharding information.
+        if hasattr(self.adapter, "sharded_state_dict"):
```
Comment (bvolpato) [inline src/megatron/bridge/peft/adapter_wrapper.py:266]: Could we preserve plain-module entries here instead of skipping them? With LoRALinear(nn.Linear, LinearAdapter), neither child implements sharded_state_dict(). I reproduced this on RTX 5070 Ti after a finite CUDA forward/backward: state_dict() contained weight, bias, adapter.linear_in.weight, and adapter.linear_out.weight, while sharded_state_dict(prefix=layer.) returned an empty dict. PEFT distributed saves start from model.sharded_state_dict() and filter afterward, so both trained LoRA matrices are already gone by then. A focused plain-linear sharded save/load test should catch this.

**Review by bvolpato — CHANGES_REQUESTED** [review]
Wrapper direction looks good, and plain `nn.Linear` behavior is well covered. I ran 151 CPU checks across changed PEFT and Moonlight files. Another 15 cases require CUDA/Transformer Engine and could not run in this Docker setup.

One checkpoint issue below is blocking. The new plain-linear wrapper can return an empty sharded state dict, which silently omits trained adapter weights.

**Review by bvolpato — CHANGES_REQUESTED** [review]
Wrapper direction looks good. All 166 targeted PEFT and Moonlight tests pass in the project CUDA 26.06 base on an RTX 5070 Ti, including 15 CUDA and Transformer Engine cases I previously could not run. A direct plain-nn.Linear LoRA CUDA forward/backward also produced finite output and loss.

Checkpoint issue below remains blocking: the same CUDA wrapper had both adapter matrices in state_dict(), but sharded_state_dict() returned an empty dict. I did not run on GB200 or run a TP=8, ETP=1, EP=64 GLM-5.2 training job.

File: src/megatron/bridge/peft/adapter_wrapper.py (hunk 4 above)
Comment (HollowMan6) [reply src/megatron/bridge/peft/adapter_wrapper.py:266]: Thank you @bvolpato ! I just fixed this and add dedicated test cases for this, please let me know if you have more comments.

**Review by cuichenx — APPROVED** [review]
LGTM

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5310/commits/fae15bb0e956204545a77ea02b7363eea31e0482

**Comment by yaoyu-33** [comment]
@HollowMan6  can you check @bvolpato 's comment see if you can follow up with a fix pr.

**Comment by HollowMan6** [comment]
@yaoyu-33 https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5310#discussion_r3718276463 It has already been fixed and tested in this PR already at https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/5310/commits/fae15bb0e956204545a77ea02b7363eea31e0482
